### PR TITLE
[SPARK-46933][SQL] Add query execution time metric to connectors which use JDBCRDD

### PR DESCRIPTION
**Pull request by urosstan-db**
### What changes were proposed in this pull request?
This pull request should add measuring query execution time on external JDBC data source.
Another change is changing access right for JDBCRDD class, that is needed for adding another metric (SQL text) which will be done in some next PR.

### Why are the changes needed?
Query execution time is very important metric to have

### Does this PR introduce _any_ user-facing change?
User can see query execution time on SparkPlan graph under node metrics tab

### How was this patch tested?
Tested using custom image

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -99,6 +99,10 @@ trait DataSourceScanExec extends LeafExecNode {
   def inputRDDs(): Seq[RDD[InternalRow]]
 }
 
+object DataSourceScanExec {
+  val numOutputRowsKey = "numOutputRows"
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:103]: This seems unnecessary change.

**Comment by cloud-fan** [comment]
Can you provide SQL web UI screenshot to demo this new metrics? And let's make sure JDBC v2 works as well (need to make `JDBCScan#supportedCustomMetrics` support this new metrics)

**Review by milastdbx — APPROVED** [review]

**Comment by urosstan-db** [comment]
> Can you provide SQL web UI screenshot to demo this new metrics? And let's make sure JDBC v2 works as well (need to make `JDBCScan#supportedCustomMetrics` support this new metrics)

image
Fetch time is removed, so it will not be present in UI. 
JDBC v2 will work with current implementation, because JDBCRelation returns JDBCScanBuilder which return JDBCScan which inherits V1Scan, so DataSourceV2Strategy will create RowDataSourceScanExec during planning.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -171,7 +173,14 @@ private[jdbc] class JDBCRDD(
...
     offset: Int)
-  extends RDD[InternalRow](sc, Nil) {
+  extends RDD[InternalRow](sc, Nil) with DataSourceMetricsMixin {
+
+  /**
+   * Execution time of the query issued to JDBC connection
+   */
+  val queryExecutionTimeMetric: SQLMetric = SQLMetrics.createNanoTimingMetric(
+    sparkContext,
+    name = "Query execution time")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:183]:
```suggestion
    name = "JDBC query execution time")
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 3]
```diff
@@ -272,11 +281,25 @@ private[jdbc] class JDBCRDD(
         ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
     stmt.setFetchSize(options.fetchSize)
     stmt.setQueryTimeout(options.queryTimeout)
+
+    val startTime = System.nanoTime
     rs = stmt.executeQuery()
+    val endTime = System.nanoTime
+
+    val executionTime = endTime - startTime
+    logInfo(s"Query execution time = $executionTime ns")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:290]: We don't need this log I think. It's redundant with web UI.

**Review by cloud-fan — APPROVED** [review]
